### PR TITLE
core, ios: separately get notification connections before requesting messages; coordinate nse threads

### PR DESCRIPTION
**Pull request by spaced4ndy**

### CODE REVIEW

File: apps/ios/SimpleX NSE/NotificationService.swift [hunk 1]
```diff
@@ -254,23 +285,42 @@ class NotificationService: UNNotificationServiceExtension {
...
+    }
+
+    func processDroppedNotifications(_ entityId: String) {
+        if !NSEThreads.shared.droppedNotifications.isEmpty {
+            let messagesToProcess = NSEThreads.shared.droppedNotifications.filter { (eId, _) in eId == entityId }
+            NSEThreads.shared.droppedNotifications.removeAll(where: { (eId, _) in eId == entityId })
+            for (index, (_, ntf)) in messagesToProcess.enumerated() {
+                if let t = threadId { logger.debug("NotificationService thread \(t, privacy: .private): entity \(entityId, privacy: .private): processing dropped notification \(index, privacy: .private)") }
+                processReceivedNtf(entityId, ntf, signalReady: false)
+            }
```
Comment (epoberezkin) [inline apps/ios/SimpleX NSE/NotificationService.swift:320]:
```suggestion
            var i: Int = 0
            var toRemove: Set<Int> = []
            for i in 0..<NSEThreads.shared.droppedNotifications.count {
              let (eId, ntf) = NSEThreads.shared.droppedNotifications[I]
              if eId == entityId {
                if let t = threadId { logger.debug("NotificationService thread \(t, privacy: .private): entity \(entityId, privacy: .private): processing dropped notification \(i, privacy: .private)") }
                processReceivedNtf(entityId, ntf, signalReady: false)
                toRemove.insert(i)
              }
            }
            NSEThreads.shared.droppedNotifications.remove(at: toRemove)
```

File: apps/ios/SimpleX NSE/NotificationService.swift [hunk 2]
```diff
@@ -280,46 +330,44 @@ class NotificationService: UNNotificationServiceExtension {
...
+            NSEThreads.shared.droppedNotifications.append((id, ntf))
+            entityReady(id, signalReady: signalReady)
+            return
         }
         if case let .msgInfo(info) = ntf {
             if info.msgId == expectedMessage.expectedMsgId {
                 logger.debug("NotificationService processNtf: msgInfo msgId = \(info.msgId, privacy: .private): expected")
                 expectedMessages[id]?.expectedMsgId = nil
-                expectedMessages[id]?.ready = true
+                entityReady(id, signalReady: signalReady)
```
Comment (epoberezkin) [inline apps/ios/SimpleX NSE/NotificationService.swift:346]:
```suggestion
                if signalReady { entityReady(id) }
```

File: apps/ios/SimpleX NSE/NotificationService.swift [hunk 3]
```diff
@@ -280,46 +330,44 @@ class NotificationService: UNNotificationServiceExtension {
...
                 self.deliverBestAttemptNtf()
-                return true
             } else if let msgTs = info.msgTs_, msgTs > expectedMsgTs {
                 logger.debug("NotificationService processNtf: msgInfo msgId = \(info.msgId, privacy: .private): unexpected msgInfo, let other instance to process it, stopping this one")
-                expectedMessages[id]?.ready = true
+                NSEThreads.shared.droppedNotifications.append((id, ntf))
+                entityReady(id, signalReady: signalReady)
```
Comment (epoberezkin) [inline apps/ios/SimpleX NSE/NotificationService.swift:351]:
```suggestion
                if signalReady { entityReady(id) }
```

File: apps/ios/SimpleX NSE/NotificationService.swift [hunk 4]
```diff
@@ -280,46 +330,44 @@ class NotificationService: UNNotificationServiceExtension {
...
-                    logger.debug("NotificationService processNtf, on apiGetConnNtfMessage: msgInfo msgId = \(info.msgId, privacy: .private), receivedMsg msgId = \(receivedMsg.msgId, privacy: .private)")
-                    return true
+                if let receivedMsg = getConnNtfMessage(connId: receiveConnId) {
+                    logger.debug("NotificationService processNtf, on getConnNtfMessage: msgInfo msgId = \(info.msgId, privacy: .private), receivedMsg msgId = \(receivedMsg.msgId, privacy: .private)")
                 } else {
-                    logger.debug("NotificationService processNtf, on apiGetConnNtfMessage: msgInfo msgId = \(info.msgId, privacy: .private): no next message, deliver best attempt")
-                    expectedMessages[id]?.ready = true
+                    logger.debug("NotificationService processNtf, on getConnNtfMessage: msgInfo msgId = \(info.msgId, privacy: .private): no next message, deliver best attempt")
+                    NSEThreads.shared.droppedNotifications.append((id, ntf))
+                    entityReady(id, signalReady: signalReady)
```
Comment (epoberezkin) [inline apps/ios/SimpleX NSE/NotificationService.swift:361]:
```suggestion
                    if signalReady { entityReady(id) }
```

File: apps/ios/SimpleX NSE/NotificationService.swift [hunk 5]
```diff
@@ -332,10 +380,21 @@ class NotificationService: UNNotificationServiceExtension {
...
             }
-            return true
+        } else {
+            NSEThreads.shared.droppedNotifications.append((id, ntf))
+            entityReady(id, signalReady: signalReady)
+        }
+    }
+
+    func entityReady(_ entityId: ChatId, signalReady: Bool) {
+        if signalReady {
```
Comment (epoberezkin) [inline apps/ios/SimpleX NSE/NotificationService.swift:390]:
```suggestion
    func entityReady(_ entityId: ChatId) {
```

File: apps/ios/SimpleX NSE/NotificationService.swift [hunk 6]
```diff
@@ -332,10 +380,21 @@ class NotificationService: UNNotificationServiceExtension {
...
+            if let t = threadId { logger.debug("NotificationService thread \(t, privacy: .private): entityReady: entity \(entityId, privacy: .private)") }
+            expectedMessages[entityId]?.ready = true
+            if let (tNext, nse) = NSEThreads.shared.activeThreads.first(where: { (_, nse) in nse.expectedMessages[entityId]?.startedProcessingNewMsgs == false }) {
+                if let t = threadId { logger.debug("NotificationService thread \(t, privacy: .private): entityReady: signal next thread \(tNext, privacy: .private) for entity \(entityId, privacy: .private)") }
+                nse.expectedMessages[entityId]?.semaphore.signal()
+            }
```
Comment (epoberezkin) [inline apps/ios/SimpleX NSE/NotificationService.swift:396]:
```suggestion
```

File: apps/ios/SimpleX NSE/NotificationService.swift (hunk 1 above)
Comment (spaced4ndy) [reply apps/ios/SimpleX NSE/NotificationService.swift:320]: reverted this refactoring:
- there is no method remove(at: <set>)
- if we do "for i in <set> { remove(at: i) }" indexes will probably get mixed up
- remove(atOffsets: ...) uses IndexSet, not Set, seems like it's usually used for UI.. probably could use it but it's untested code

**Pull request merged**